website: Added documentation for using local datasources in gem5 (#90)
gem5/new-website: _pages/documentation/gem5-stdlib/7-using-local-data-sources.md

Reviewer: I think for this document to be complete you'll need to explain how one would setup their own MongoDB atlas instance and the schema of it. While it's obvious to us, it won't be to the users that the JSON file and the MongoDB share the same schema.

---
layout: documentation
title: Using local Data sources in gem5
parent: gem5-standard-library
doc: gem5 documentation
permalink: /documentation/gem5-stdlib/using-local-resources
author: Harshil Patel
---

gem5 supports using local data sources in the form of a MongoDB Atlas and JSON datasource. To utilize data sources other than the main gem5 resources database, you will need to override the gem5-resources-config.

> MongoDB Config Format:

Reviewer 2: There are a few things that need more explanation.
- Each field of the MongoDB format and the raw JSON file format should have a brief description.
- Since the MongoDB JSON and the raw JSON files have different amount of information, we'd need to explain why or when to use what.

```json
{
    "sources":{
        "example-atlas-config": {
            "dataSource": "datasource name",
            "database": "database name",
            "collection": "collection name",
            "url": "Atlas data API URL",
            "authUrl": "Atlas authentication URL",
            "apiKey": "API key",
            "isMongo": true
        }       
    }
}
```
JSON Config Format:

```json
{
    "sources":{
        "example-json-config": {
            "url": "local path to JSON file or URL to a JSON file",
            "isMongo": false
        }
    }
}
```

> ### Updating gem5 Resources Config:

Reviewer 2: Nit:

```suggestion
### Updating gem5 Resources Config
```

> You can update the gem5 resources config in the following ways:

Reviewer 2: I think we should rephrase L42, L44, L46, and L48. The problem is that it is unclear what `GEM5_CONFIG`, `GEM5_RESOURCE_JSON`, and `GEM5_RESOURCE_JSON_APPEND`. It is also unclear what happens when we modify more than one of the mentioned variables at the same time. Also, since there are many ways of modifying the config, is there a preferable way of doing this? I think we should have fewer ways of modifying the configs.

Set the GEM5_CONFIG environment variable to point to a new config file. This will override the default config.

Have a file named gem5-config.json in the current working directory. This will also override the default config.

Use the GEM5_RESOURCE_JSON environment variable to override the default config to use a JSON file, or use the GEM5_RESOURCE_JSON_APPEND environment variable to add a JSON file to the current config. If the config is overridden by the above two methods, then these environment variables will override or modify that config.

### Using Multiple Data Sources

> gem5 supports the use of more than one data source. The config structure is as follows:

Reviewer 2: We should use "resource configuration" to refer to the json file containing the resource metadata. Saying "configuration" alone is confusing because we also use "configuration" in many other places, like "python system configuration" for the python file inputted to gem5.

```suggestion
gem5 supports the use of more than one data source. The structure of resource configuration is as follows:
```

```json
{
    "sources": {
        "data-source-1": { ... },
        "data-source-2": { ... },
        // Add more data sources as needed
    }
}
```

The data sources can be a combination of MongoDB Atlas and JSON. By default, gem5 will search through all the data sources to find a resource and return the latest version of the resource compatible with the gem5 version if the `resource_version` is not given. If a resource with the same `id` and `resource_version` exists in multiple data sources, an error will be thrown. You can also specify a subset of data sources to obtain resources from:

```python
resource = obtain_resource("id", clients=["data-source-1"])
```

> ### Using Custom Config and Local Resources

Reviewer 2: ```suggestion
### Using Custom Resource Configuration and Local Resources
```

        "code_examples": [
            {
                "example": "https://github.com/gem5/gem5/tree/develop/configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py",
                "tested": true
            },
            {
                "example": "https://github.com/gem5/gem5/tree/develop/configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py",
                "tested": true
            },
            {
                "example": "https://github.com/gem5/gem5/tree/develop/configs/example/gem5_library/riscvmatched-hello.py",
                "tested": true
            }
        ],
        "license": "",
        "author": [],
        "source_url": "https://github.com/gem5/gem5-resources/tree/develop/src/simple",
        "resource_version": "1.0.0",
        "gem5_versions": [
            "23.0"
        ],
        "example_usage": "obtain_resource(resource_id=\"riscv-hello\")"
    }
]
```

The JSON object of a resource should adhere to the [gem5 resources schema](https://resources.gem5.org/gem5-resources-schema.json). 

**Note**: While the `url` field can be a link, in this case, I'm using a local download.

> #### Creating Your Custom Configuration

Reviewer 2: ```suggestion
#### Creating Your Custom Resource Configuration
```

---
layout: documentation

> title: Using local Data sources in gem5

Reviewer: Let's work on our terminology here. I realize there's no standardization here but i think 'Data source" is really broad. How is someone supposed to know this is refering to the gem5 Resources "Data source".

My suggestion would be "Setting gem5 Resources data sources".

parent: gem5-standard-library
doc: gem5 documentation

> permalink: /documentation/gem5-stdlib/using-local-resources

Reviewer: Just adding this file with this permalink isn't enough. You need to update "_data/documentation.yml" for this to be building into the website.

I'd strongly recommend figuring out how to build the and host the website locally to verify website changes.

author: Harshil Patel
---

> gem5 supports using local data sources in the form of a MongoDB Atlas and JSON datasource. To utilize data sources other than the main gem5 resources database, you will need to override the gem5-resources-config.

Reviewer: I think you should lead in with an explanation/reminder that gem5 resources references a data source in order to get data about resources and, by default this is a MongoDB database. Then say the user may set local data source.

I think the way you phrase it should state that the database we use is also MondoDB. The phrasing here is close to suggesting local data sources can be mongoDB or JSON and the main gem5 resources is some mysterous other data source type

Reviewer: "you will need to override the gem5-resources-config"

The user likely has no idea 1) what this is and, 2) where it is.

Please take a moment to explain it's purpose, where it is, and the ways of modifying/creating it (via environment variable or setting config.json in the gem5 the CWD). Tell them where the default values come from in addition. It'll make things clearer to them.

((Edit: I see you explain this after the examples. I'd move much of this to be before the examples))

> gem5 supports using local data sources in the form of a MongoDB Atlas and JSON datasource. To utilize data sources other than the main gem5 resources database, you will need to override the gem5-resources-config.

Reviewer: One more thing: Calling these "local" isn't right. We don't mean local. If they hose their own MongoDB instance on the cloud, that's not local. I'd just phrase things in the style of "Add a gem5 Resources MongoDB Datasource".

            "isMongo": true
        }       
    }
}
```
JSON Config Format:

```json
{
    "sources":{
        "example-json-config": {
            "url": "local path to JSON file or URL to a JSON file",
            "isMongo": false
        }
    }
}
```
### Updating gem5 Resources Config

You can update the gem5 resources config in the following ways:

 - Set the GEM5_CONFIG environment variable to point to a new config file. This will override the default config.

- Have a file named gem5-config.json in the current working directory. This will also override the default config.

- If the above two methods are not used, then the default resources config will be used.

You can also override the resources config with a JSON file or add a JSON file to the config as follows:

- Use the GEM5_RESOURCE_JSON environment variable to override the default config to use a JSON file.

- GEM5_RESOURCE_JSON_APPEND environment variable to add a JSON file to the current config. 

If the gem5 resources config was updated by the GEM5_CONFIG environment variable or having a gem5-config.json in current working directory, then these flags will override or add to those configs.

> 

Reviewer: I think this is confusing. At least it was to me, and i'm supposed to know this. As a side note: I'm starting to regret supporting both these cases because it's getting messy with all this "if this then that but not in the case of x where then default z is overridden".

To aid understanability I'd create two completely seperate sections "Add a MongoDB data sources" and "Add a JSON file data source".

Author: The above three bullet points (points before line 50)  works with both JSON and MongoDB database or a combination of them but the later two are specifically for one JSON. I can clarify the use case for the later points is that you have a local JSON which you want to use to run gem5.

Reviewer: Ah ok. if you can, just try to separate these things off into some logical sections. I realize there's overlap here and I trust you to handle it. I just don't want to confuse users by talking about two things at once or jumping to face between explaining the two separate data sources.

Note: Overriding or appending does not modify the actual config files.

### Using Multiple Data Sources

gem5 supports the use of more than one data source. The structure of the resource configuration is as follows:

```json
{
    "sources": {
        "data-source-1": { ... },
        "data-source-2": { ... },
        // Add more data sources as needed
    }
}
```

> The data sources can be a combination of MongoDB Atlas and JSON. By default, gem5 will search through all the data sources to find a resource and return the latest version of the resource compatible with the gem5 version if the `resource_version` is not given. If a resource with the same `id` and `resource_version` exists in multiple data sources, an error will be thrown. You can also specify a subset of data sources to obtain resources from:

Reviewer: I'd drop the first sentense. Just show in the example the sources containing 1 mongodb and 2 JSON files. Then add a sentence below saying the above example shows a resource sources configuraiton consisting of 1 MongoDB Atlas data source and two JSON file sources. 

The second sentence (By default...) isn't needed, it goes too into details that aren't relevant to understanding this. What i think needs said is "gem5 will create a union of all the gem5-resources data entries across all the data sources".

```python
resource = obtain_resource("id", clients=["data-source-1"])
```

> ### Using Custom Resource Configuration and Local Resources

Reviewer: You jumped straight into this without explaining Local Resources at all. I'd explain what they are or at least provide a sentence and a link to where they can learn more.

I'd also explain the use case here. Why would a user want to do this? What are they trying to do. Paint them a picture.

In this example, we will walk through how to set up your custom configuration and utilize your own local resources. For this illustration, we'll employ a JSON file as our resource data source.

#### Creating a Custom Resource Data Source

Let's begin by downloading the resource we'll use as our local source. I'm selecting the `riscv-hello` resource from gem5 resources.

```bash
wget https://storage.googleapis.com/dist.gem5.org/dist/develop/test-progs/hello/bin/riscv/linux/hello-20220728

> ```

Reviewer: I can see this confusing some people. Just have them create a text file called "resourde-a.txt" or something like that. It's not important.

        "size": 4674040,
        "tags": [
            "hello"
        ],
        "is_zipped": false,
        "md5sum": "6d9494d22b90d817e826b0d762fda973",
        "source": "src/simple",
        "url": "file://<path/to/the/above/downloaded/file>",
        "code_examples": [
            {
                "example": "https://github.com/gem5/gem5/tree/develop/configs/example/gem5_library/checkpoints/riscv-hello-restore-checkpoint.py",
                "tested": true
            },
            {
                "example": "https://github.com/gem5/gem5/tree/develop/configs/example/gem5_library/checkpoints/riscv-hello-save-checkpoint.py",
                "tested": true
            },
            {
                "example": "https://github.com/gem5/gem5/tree/develop/configs/example/gem5_library/riscvmatched-hello.py",
                "tested": true
            }
        ],
        "license": "",
        "author": [],
        "source_url": "https://github.com/gem5/gem5-resources/tree/develop/src/simple",
        "resource_version": "1.0.0",
        "gem5_versions": [
            "23.0"
        ],
        "example_usage": "obtain_resource(resource_id=\"riscv-hello\")"
    }
]
```

> The JSON object of a resource should adhere to the [gem5 resources schema](https://resources.gem5.org/gem5-resources-schema.json). 

Reviewer: This is probably the best w can do for now, but this is going to completely block users from proceeding. It's far too hard to determine the schema from this.

Maybe we need to provide some example JSON with comments and documentation for each category?

**Note**: While the `url` field can be a link, in this case, I'm using a local download.

#### Creating Your Custom Resource Configuration

Create a file named `gem5-config.json` with the following content:

```json
{
    "sources": {
        "my-json-data-source": {

>             "url": "path/to/my-resources.json",

Reviewer: Doesn't this have to be `file://path/to/my-resources.json`?

>             "isMongo": false

Reviewer: Worth explaining this very briefly: I.e., `(isMongo == True) == (isJSON == False)` :)

        }
    }
}
```

#### Running gem5 with a Local Data Source

First, build gem5 with RISC-V:

```bash
scons build/RISCV/gem5.opt -j`nproc`
```

Next, run the `riscvmatched-hello.py` example using our local `riscv-hello` resource:

```bash
GEM5_CONFIG=path/to/gem5-config.json ./build/RISCV/gem5.opt configs/example/gem5_library/riscvmatched-hello.py
```

> This command will execute the `riscvmatched-hello.py` script using our locally downloaded resource.

Reviewer: I'd create something more fake here. Just have them write a config that calls `obtain_resource` then print off some data from the resource like the local path. Then mention this returned resource can be used in a simulation, just like those obtained from the official gem5 MongoDB datasource.

    }
}
```
### Updating gem5 Resources Config

You can update the gem5 resources config in the following ways:

>  - Set the GEM5_CONFIG environment variable to point to a new config file. This will override the default config.

Reviewer 3: What is the "default config"? Can you define this, please.
